pkg/cloudprovider/providers/vsphere/vsphere_util.go

```diff
@@ -63,3 +81,226 @@ func GetgovmomiClient(cfg *VSphereConfig) (*govmomi.Client, error) {
+	compatibleHubs := compatibilityResult.CompatibleDatastores()
+	// Return an error if there are no compatible datastores.
+	if len(compatibleHubs) < 1 {
+		return nil, fmt.Errorf("There are no compatible datastores that satisfy the storage policy requirements")
+	}
+	dsMoList, err := vs.getDatastoreMo(ctx, compatibleHubs)
+	if err != nil {
+		return nil, err
+	}
+	return dsMoList, nil
+}
+
+func (vs *VSphere) GetNonCompatibleDatastoresMo(ctx context.Context, compatibilityResult pbm.PlacementCompatibilityResult) []mo.Datastore {
+	nonCompatibleHubs := compatibilityResult.NonCompatibleDatastores()
+	// Return an error if there are no compatible datastores.
+	if len(nonCompatibleHubs) < 1 {
+		return nil
+	}
+	dsMoList, err := vs.getDatastoreMo(ctx, nonCompatibleHubs)
+	if err != nil {
+		return nil
+	}
+	return dsMoList
+}
+
+// Get the datastore managed objects for the place hubs using property collector.
+func (vs *VSphere) getDatastoreMo(ctx context.Context, hubs []pbmtypes.PbmPlacementHub) ([]mo.Datastore, error) {
+	var dsMoRefs []types.ManagedObjectReference
+	for _, hub := range hubs {
+		dsMoRefs = append(dsMoRefs, types.ManagedObjectReference{
+			Type:  hub.HubType,
+			Value: hub.HubId,
+		})
+	}
+
+	pc := property.DefaultCollector(vs.client.Client)
+	var dsMoList []mo.Datastore
+	err := pc.Retrieve(ctx, dsMoRefs, []string{DatastoreInfoProperty}, &dsMoList)
+	if err != nil {
+		return nil, err
+	}
+	return dsMoList, nil
+}
+
+// Get all datastores accessible for the virtual machine object.
+func (vs *VSphere) getSharedDatastoresInK8SCluster(ctx context.Context) ([]types.ManagedObjectReference, error) {
+	f := find.NewFinder(vs.client.Client, true)
+	dc, err := f.Datacenter(ctx, vs.cfg.Global.Datacenter)
+	f.SetDatacenter(dc)
+	vmFolder, err := f.Folder(ctx, strings.TrimSuffix(vs.cfg.Global.WorkingDir, "/"))
+	if err != nil {
+		return nil, err
+	}
+	vmMoList, err := vs.GetVMsInsideFolder(ctx, vmFolder, []string{NameProperty})
+	if err != nil {
+		return nil, err
+	}
+	index := 0
+	var sharedDs []string
+	for _, vmMo := range vmMoList {
+		if !strings.HasPrefix(vmMo.Name, DummyVMPrefixName) {
+			accessibleDatastores, err := vs.getAllAccessibleDatastores(ctx, vmMo)
+			if err != nil {
+				return nil, err
+			}
+			if index == 0 {
+				sharedDs = accessibleDatastores
+			} else {
+				sharedDs = intersect(sharedDs, accessibleDatastores)
+				if len(sharedDs) == 0 {
+					return nil, fmt.Errorf("No shared datastores found in the Kubernetes cluster")
+				}
+			}
+			index++
+		}
+	}
+	var sharedDSMorefs []types.ManagedObjectReference
+	for _, ds := range sharedDs {
+		sharedDSMorefs = append(sharedDSMorefs, types.ManagedObjectReference{
+			Value: ds,
+			Type:  "Datastore",
+		})
+	}
+	return sharedDSMorefs, nil
+}
+
+func intersect(list1 []string, list2 []string) []string {
+	var sharedList []string
+	for _, val1 := range list1 {
```

Comment:
How big do you expect this list to be?  You could potentially do this is O(n) instead of O(n^2) time.